File: Stacks/Stacks/Models/Badge.swift
import Foundation

struct Badge: Codable, Identifiable {
    let id: Int
    let name: String
    let tier: BadgeTier
    let iconUrl: String?
    let criteria: String?
    let earnedAt: Date?
    
    enum CodingKeys: String, CodingKey {
        case id
        case name
        case tier
        case iconUrl
        case criteria
        case earnedAt
    }
}

enum BadgeTier: String, Codable {
    case bronze
    case silver
    case gold
    case platinum
}

File: backend/src/entities/Badge.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToMany,
} from 'typeorm';
import { UserBadge } from './UserBadge';

export enum BadgeTier {
  BRONZE = 'bronze',
  SILVER = 'silver',
  GOLD = 'gold',
  PLATINUM = 'platinum',
}

@Entity('badges')
export class Badge {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column()
  name!: string;

  @Column({
    type: 'enum',
    enum: BadgeTier,
  })
  tier!: BadgeTier;

  @Column({ name: 'icon_url', nullable: true })
  iconUrl?: string;

  @Column({ type: 'text', nullable: true })
  criteria?: string;

  @OneToMany(() => UserBadge, (userBadge) => userBadge.badge)
  userBadges!: UserBadge[];
}

